File: 0001-0050/0019/solution.cpp
#include <iostream>
#include <vector>

using namespace std;

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode *dummy = new ListNode(0);
        ListNode *slow = dummy, *fast = dummy;
        dummy->next = head;
        for (int i = 0; i < n; i++) {
            fast = fast->next;
        }
        while (fast->next) {
            slow = slow->next;
            fast = fast->next;
        }
        slow->next = slow->next->next;
        return dummy->next;
    }
};

struct Example {
    ListNode *head;
    int n;
    ListNode *ans;
};

ListNode* buildList(vector<int> nums) {
    ListNode *dummy = new ListNode(0), *cur = dummy;
    for (auto& num: nums) {
        cur->next = new ListNode(num);
        cur = cur->next;
    }
    return dummy->next;
}

bool check(ListNode* l1, ListNode* l2) {
    ListNode *p = l1, *q = l2;
    while (p && q && p->val == q->val) {
        p = p->next;
        q = q->next;
    }
    return p == q;
}

int main() {
    vector<Example> examples;
    examples.push_back({
        buildList({1, 2, 3, 4, 5}),
        2,
        buildList({1, 2, 3, 5}),
    });
    examples.push_back({
        buildList({1}),
        1,
        buildList({}),
    });
    examples.push_back({
        buildList({1, 2}),
        1,
        buildList({1}),
    });

    
    Solution *solve = new Solution();
    for (int i = 0; i < examples.size(); i++) {
        ListNode *ans = solve->removeNthFromEnd(examples[i].head, examples[i].n);
        if (check(ans, examples[i].ans)) {
            cout << "PASS: CASE " << i << endl;
        } else {
            cout << "FAIL: CASE " << i << endl;
        }
    }
    
    return 0;
}

File: 0001-0050/0019/solution.py
from typing import Optional


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        dummy = ListNode(0)
        slow, fast = dummy, dummy
        dummy.next = head
        for i in range(n):
            fast = fast.next
        while fast.next:
            slow = slow.next
            fast = fast.next
        slow.next = slow.next.next
        return dummy.next


class Example:
    def __init__(self, head, n, ans):
        self.head = head
        self.n = n
        self.ans = ans


def buildList(nums):
    dummy = ListNode(0)
    cur = dummy
    for num in nums:
        cur.next = ListNode(num)
        cur = cur.next
    return dummy.next


def check(l1, l2):
    p, q = l1, l2
    while p and q and p.val == q.val:
        p = p.next
        q = q.next
    return p == q


def main():
    examples = [Example(
        buildList([1, 2, 3, 4, 5]),
        2,
        buildList([1, 2, 3, 5])
    ), Example(
        buildList([1]),
        1,
        buildList([])
    ), Example(
        buildList([1, 2]),
        1,
        buildList([1])
    )]

    solve = Solution()
    for i, example in enumerate(examples):
        ans = solve.removeNthFromEnd(example.head, example.n)
        if check(ans, example.ans):
            print(f"PASS: CASE {i}")
        else:
            print(f"FAIL: CASE {i}")


if __name__ == "__main__":
    main()
